lib/linguist/heuristics.yml

```diff
@@ -48,7 +48,13 @@ disambiguations:
 - extensions: ['.as']
   rules:
   - language: ActionScript
-    pattern: '^\s*(package\s+[a-z0-9_\.]+|import\s+[a-zA-Z0-9_\.]+;|class\s+[A-Za-z0-9_]+\s+extends\s+[A-Za-z0-9_]+)'
+    pattern: 
+    - '^\s*package(?:\s+[\w\.]+)?\s+[{\n]{1}'
+    - '^\s*import\s+[\w\.\*]+\s*;'
+    - '^\s*intrinsic\s+class\s+[\w<>\.]+'
+    - '^\s*class\s+[\w<>\.]+\s+extends\s+[\w<>\.]+'
+    - '^\s*(?:(?:public|protected|private|static|var|const|local)\s+)+\w+\s*:\s*[\w<>\.]+(?:\s*=.*)?\s*;'
+    - '^\s*(?:(?:public|protected|private|static)\s+)*function\s+[\w]+\s*\((?:\s*\w+(?:\s*:\s*[\w<>\.]+)?,?)*\)(?:\s*:\s*[\w<>\.]+)?'
```

Comment:
Here's a more optimal expression...

~~~yaml
pattern: '^\s*(?:package(?:\s+[\w.]+)?\s+(?:{|$)|import\s+[\w.*]+\s*;|(intrinsic\s+)?class\s+[\w<>.]+(?:\s+extends\s+[\w<>.]+)?|(?:(?:public|protected|private|static)\s+)?(?:(?:var|const|local)\s+\w+\s*:\s*[\w<>.]+(?:\s*=.*)?\s*;|function\s+\w+\s*\((?:\s*\w+(?:\s*:\s*[\w<>.]+)?,?)*\)(?:\s*:\s*[\w<>.]+)?))'
~~~

... and the readable version of it is:

~~~regexp
^\s*
(?:
	package (?:\s+[\w.]+)? \s+ (?:{|$)
	|
	import \s+ [\w.*]+ \s* ;
	|
	(intrinsic\s+)? class \s+ [\w<>.]+ (?:\s+extends\s+[\w<>.]+)?
	|
	(?:(?:public|protected|private|static)\s+)?
	(?:
		# Variable
		(?:var|const|local) \s+ \w+ \s*        # var foo
		: \s* [\w<>.]+                         # : Type
		(?:\s*=.*)? \s* ;                      # = ... ;
		
		|
		
		# Function
		function \s+ \w+ \s*                   # function fooBar
		\((?:\s*\w+(?:\s*:\s*[\w<>.]+)?,?)*\)  # (arg1, arg2)
		(?: \s* : \s* [\w<>.]+ )?              # :<returnValue>
	)
)
~~~

Reply:
@Ryan-rsm-McKenzie Righto, I've updated the pattern.